packages/jsii-pacmak/lib/targets/python/type-name.ts

```diff
@@ -0,0 +1,354 @@
+import {
+  Assembly,
+  OptionalValue, TypeReference,
+  isCollectionTypeReference, CollectionKind,
+  isNamedTypeReference,
+  isPrimitiveTypeReference, PrimitiveType, PrimitiveTypeReference,
+  isUnionTypeReference,
+  Type,
+} from '@jsii/spec';
+import { createHash } from 'crypto';
+import { die, toPythonIdentifier } from './util';
+
+export interface TypeName {
+  pythonType(context: NamingContext): string;
+  requiredImports(context: NamingContext): PythonImports;
+}
+
+export interface PythonImports {
+  /**
+   * For a given source module, what elements to import. The empty string value
+   * indicates a need to import the module fully ("import <name>") instead of
+   * doing a piecemeal import ("from <name> import <item>").
+   */
+  readonly [sourcePackage: string]: ReadonlySet<string>;
+}
+
+/**
+ * The context in which a PythonType is being considered.
+ */
+export interface NamingContext {
+  /** The assembly in which the PythonType is expressed. */
+  readonly assembly: Assembly;
+
+  /** The submodule of the assembly in which the PythonType is expressed (could be the module root) */
+  readonly submodule: string;
+
+  /**
+   * The declaration is made in the context of a type annotation (so it can be quoted)
+   *
+   * @default true
+   */
+  readonly typeAnnotation?: boolean;
+
+  /**
+   * The nesting scope in which the PythonType is expressed (if any)
+   *
+   * @default - none
+   */
+  readonly nestingScope?: string | undefined;
+
+  /**
+   * Disables generating typing.Optional wrappers
+   * @default false
+   * @internal
+   */
+  readonly ignoreOptional?: boolean;
+}
+
+export function toTypeName(ref?: OptionalValue | TypeReference): TypeName {
+  if (ref == null) {
+    return Primitive.NONE;
+  }
+
+  const type = isOptionalValue(ref) ? ref.type : ref;
+  const optional = isOptionalValue(ref) && ref.optional;
```

Comment:
doesn't `isOptionalValue` imply `ref.optional === true`?

Reply:
No - `isOptionalValue(x) -> x is OptionalValue`. I'd agree the naming is a little awkward here (but at the time I didn't think of anything better 🤷🏻‍♂️)

Reply:
Maybe `isOptionalValueType`?